File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.cpp
#include <string>
#include <vector>
#include <cmath>
#include <iostream>
using namespace std;

string solution(vector<int> numbers, string hand) {
    string answer = "";

    int coor[12][2] = {{3,1},{0, 0},{0, 1},{0, 2},{1, 0},{1,1},{1,2},{2,0},{2,1},{2,2},{3,0},{3,2}};
    int dist[12][12] = {0};
    
    for (int i =0 ; i < 12;i ++){
        for (int j =0; j < 12;j ++)
            dist[i][j] = abs(coor[i][0]-coor[j][0]) + abs(coor[i][1]-coor[j][1]);
    }
    
    int preL = 10, preR = 11;
    
    for (int n : numbers){
        char now = hand == "right" ? 'R' : 'L';
            
        if (n == 1 || n == 4 || n == 7)
            now = 'L';
        else if (n == 3 || n == 6 || n == 9)
            now = 'R';
        else{
            if (dist[preL][n] != dist[preR][n]){
                if (dist[preL][n] < dist[preR][n])
                    now = 'L';
                else
                    now = 'R';
            }
        }
        
        answer += now;
        
        if (now == 'L')
            preL = n;
        else
            preR = n;
    }
    
    return answer;
}

File: 프로그래머스/1/67256. ［카카오 인턴］ 키패드 누르기/［카카오 인턴］ 키패드 누르기.py
def solution(numbers, hand):
    core = {i: [(i-1) // 3, (i-1) % 3] for i in range(1, 10)}
    core[0] = [3, 1]
    
    l = [3, 0]
    r = [3, 2]
    ans = ''
    
    for num in numbers:
        if num in (1, 4, 7):
            l = core[num]
            ans += 'L'
        elif num in (3, 6, 9):
            r = core[num]
            ans += 'R'
        else:
            Ldist = abs(core[num][0] - l[0]) + abs(core[num][1] - l[1])
            Rdist = abs(core[num][0] - r[0]) + abs(core[num][1] - r[1])
            
            if Ldist < Rdist:
                l = core[num]
                ans += 'L'
            elif Ldist > Rdist:
                r = core[num]
                ans += 'R'
            else:
                if hand == 'right':
                    r = core[num]
                    ans += 'R'
                else:
                    l = core[num]
                    ans += 'L'
                
    return ans